demos/inpage/src/accounts/SafeECDSAAccountWrapper.ts

```diff
@@ -0,0 +1,138 @@
+    public waxInPage: WaxInPage,
+  ) {}
+
+  static fromData(data: SafeECDSAAccountData, waxInPage: WaxInPage) {
+    return new SafeECDSAAccountWrapper(
+      data.address,
+      data.privateKey,
+      data.ownerAddress,
+      waxInPage,
+    );
+  }
+
+  toData(): SafeECDSAAccountData {
+    return {
+      type: 'SafeECDSAAccount',
+      address: this.address,
+      privateKey: this.privateKey,
+      ownerAddress: this.ownerAddress,
+    };
+  }
+
+  static async createRandom(
+    waxInPage: WaxInPage,
+  ): Promise<SafeECDSAAccountWrapper> {
+    const contracts = await waxInPage.getContracts();
+
+    const wallet = ethers.Wallet.createRandom();
+
+    const admin = await waxInPage.requestAdminAccount('deploy-account');
+
+    const createArgs = [
+      contracts.safe,
+      contracts.entryPoint,
+      wallet,
+      0,
+    ] satisfies Parameters<SafeECDSAFactory['create']>;
+
+    const address = await contracts.safeECDSAFactory.create.staticCall(
+      ...createArgs,
+    );
+
+    await receiptOf(
+      contracts.safeECDSAFactory.connect(admin).create(...createArgs),
+    );
+
+    return new SafeECDSAAccountWrapper(
+      address,
+      wallet.privateKey,
+      wallet.address,
+      waxInPage,
+    );
+  }
+
+  getContract(): SafeECDSAPlugin {
+    return SafeECDSAPlugin__factory.connect(
+      this.address,
+      this.waxInPage.ethersProvider,
+    );
+  }
+
+  // eslint-disable-next-line class-methods-use-this, @typescript-eslint/require-await
+  async makeInitCode(): Promise<string> {
+    throw new Error(
+      'SafeECDSAAccount does not use initCode (it must be created before use)',
+    );
+  }
+
+  // eslint-disable-next-line @typescript-eslint/require-await, class-methods-use-this
+  async encodeActions(actions: EthereumRpc.Action[]): Promise<string> {
+    if (actions.length !== 1) {
+      throw new Error('SafeECDSAAccount only supports single actions');
+    }
+
+    const action = actions[0];
+
+    return SafeECDSAPlugin__factory.createInterface().encodeFunctionData(
+      'execTransaction',
+      [action.to, action.value ?? 0n, action.data ?? '0x'],
+    );
+  }
+
+  // eslint-disable-next-line class-methods-use-this, @typescript-eslint/require-await
+  async estimateVerificationGas(
+    _userOp: EthereumRpc.UserOperation,
+  ): Promise<bigint> {
+    // TODO: estimateGas on validateUserOp?
+    return baseVerificationGas;
+  }
+
+  async getNonce(): Promise<bigint> {
+    const contracts = await this.waxInPage.getContracts();
+
+    // TODO: Why does this give a different result to
+    // this.getContract().getNonce()?
+    // (And why does that alternative give the wrong answer?)
+    return await contracts.entryPoint.getNonce(this.address, 0);
+  }
+
+  async sign(
+    _userOp: EthereumRpc.UserOperation,
+    userOpHash: string,
+  ): Promise<string> {
+    const ownerWallet = new ethers.Wallet(this.privateKey);
+
+    return await ownerWallet.signMessage(ethers.getBytes(userOpHash));
```

Comment:
Should verify `_userOp` corresponds to signed userOpHash?

Reply:
Hmm. I wrote it this way because it's an internal tool and `userOpHash` is already available at the call site, so it's easier to just provide it rather than recalculating it.

Having said that, the call site looks like this:

```ts
      let userOpHash = await contracts.entryPoint.getUserOpHash(userOp);

      userOp.signature = await account.sign(userOp, userOpHash);
```

This means that the RPC node could get us to sign anything by lying about the `userOpHash`. We generally do put a lot of trust in the RPC node, but calculating the `userOpHash` locally is probably a good idea.

Reply:
Fixed: https://github.com/getwax/wax/pull/100